facet-reflect/src/peek/fields.rs

```diff
@@ -0,0 +1,210 @@
+
+        Self {
+            range: 0..fields.len(),
+            state: FieldIterState::Enum {
+                peek_enum: enum_,
+                fields,
+            },
+        }
+    }
+
+    pub(crate) fn new_tuple(tuple: PeekTuple<'mem, 'facet, 'shape>) -> Self {
+        Self {
+            range: 0..tuple.len(),
+            state: FieldIterState::Tuple(tuple),
+        }
+    }
+
+    fn get_field_by_index(
+        &self,
+        index: usize,
+    ) -> Option<(Field<'shape>, Peek<'mem, 'facet, 'shape>)> {
+        match self.state {
+            FieldIterState::Struct(peek_struct) => {
+                let field = peek_struct.ty.fields.get(index).copied()?;
+                let value = peek_struct.field(index).ok()?;
+                Some((field, value))
+            }
+            FieldIterState::Tuple(peek_tuple) => {
+                let field = peek_tuple.ty.fields.get(index).copied()?;
+                let value = peek_tuple.field(index)?;
+                Some((field, value))
+            }
+            FieldIterState::Enum { peek_enum, fields } => {
+                // Get the field definition
+                let field = fields[index];
+                // Get the field value
+                let field_value = match peek_enum.field(index) {
+                    Ok(Some(v)) => v,
+                    Ok(None) => return None,
+                    Err(e) => panic!("Cannot get field: {:?}", e),
+                };
+                // Return the field definition and value
+                Some((field, field_value))
+            }
+            FieldIterState::FlattenedEnum { field, value } => {
+                if index == 0 {
+                    Some((field, value))
+                } else {
+                    None
+                }
+            }
+        }
+    }
+}
+
+impl<'mem, 'facet, 'shape> Iterator for FieldIter<'mem, 'facet, 'shape> {
+    type Item = (Field<'shape>, Peek<'mem, 'facet, 'shape>);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let index = self.range.next()?;
+
+            let Some(field) = self.get_field_by_index(index) else {
+                continue;
+            };
+
+            return Some(field);
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.range.size_hint()
+    }
+}
+
+impl DoubleEndedIterator for FieldIter<'_, '_, '_> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        loop {
+            let index = self.range.next_back()?;
+
+            let Some(field) = self.get_field_by_index(index) else {
+                continue;
+            };
+
+            return Some(field);
+        }
+    }
+}
+
+impl ExactSizeIterator for FieldIter<'_, '_, '_> {}
+
+/// An iterator over the fields of a struct or enum that should be serialized. See [`HasFields::fields_for_serialize`]
+pub struct FieldsForSerializeIter<'mem, 'facet, 'shape> {
+    stack: Vec<FieldIter<'mem, 'facet, 'shape>>,
+}
```

Comment:
This type is a hand-written iterator to re-create the old `fields_for_serialize()` impl Trait method. This was necessary so that `facet-serialize` could store a nameable iterator type (I also tried `Box<dyn Iterator<...>>` but hit some weird lifetime bounds issues... plus I didn't feel great about needing an extra box for this

(One minor plus to this implementation is that it's no longer recursive!)